staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go

```diff
@@ -223,6 +223,16 @@ func WithTypeAndObjectMeta(s *schema.Structural) *schema.Structural {
 	return result
 }
 
+// MaxCardinality returns the maximum number of times data conforming to the schema could possibly exist in
+// an object serialized to JSON. For cases where a schema is contained under map or array schemas of unbounded
+// size, this can be used as an estimate as the worst case number of times data matching the schema could be repeated.
+// Note that this only assumes a single comma between data elements, so if the schema is contained under only maps,
+// this estimates a higher cardinality that would be possible.
+func MaxCardinality(s *schema.Structural) uint64 {
+	sz := estimateMinSizeJSON(s) + 1 // assume at least one comma between elements
```

Comment:
is this being called in ways that will make us repeatedly recursively evaluate the size of a schema?

if I have a schema 40 nesting levels deep, and have a cel rule at each level, does this call:
* estimateMinSizeJSON(root) (traversing all child schemas to compute the min size of the root)
* estimateMinSizeJSON(level 1) (re-traversing all child schemas to compute the min size of level 1)
* ...
?

Reply:
beyond the scope of this PR, but a similar question exists for other callers of estimateMinSizeJSON via SchemaDeclType / estimateMaxArrayItemsPerRequest / estimateMaxAdditionalPropertiesPerRequest

we want to make sure a deep schema with cel rules at the root and other levels isn't going to be super expensive to compute cost on

Reply:
This general problem is fairly pervasive in the CRD validation side of things. While I was looking at how we could do more work as a post traversal step to make the min calculations cheap, I noticed that we construct a new structural schema whenever we compile CEL programs, which is another case of use doing a recursive traversal at every level (for the worst case). We also convert those schemas to the "decl" format that CEL accepts in compile (which is another recursive traversal).

How would you feel about a beta task where we construct a benchmark that reproduces this problem well and then optimize it away?  A traversal that starts at the first branch in the tree where a CEL rule encountered that accumulates some base facts (like min sizes), and prepares the structural and "decls" schemas should allow for a lot of reuse.  But it's a larger change.

Reply:
sounds ok for beta, as long as the calculations that do recursive traversals in this PR and #108990 are behind the cel validation feature gate